src/Discord.Net/IDiscordClient.cs

```diff
@@ -0,0 +1,28 @@
+using System;
+using System.Threading.Tasks;
+using Wumpus;
+
+namespace Discord
+{
+    public interface IDiscordClient
+    {
+        event Action Connected;
+        event Action Disconnected;
+        event Action Ready;
+
+        event Action<IChannel> ChannelCreated;
```

Comment:
Are events no longer asynchronous? If not, presumably we become responsible for dispatching the event to a new `async` context, if desired?

I'm totally cool with that, since that's what we do currently anyway, but it might be a good idea to have a built-in mechanism to do said dispatching, for beginner consumers. Something like an injectable `IDiscordEventDispatcher`, with a couple built-in implementations?

Reply:
In my mind, the ideal way to handle discord events is the same way ASP.NET handles HTTP requests: each incoming message spawns a new "logical" thread. I.E. a new `IServiceScope` in the DI system, and a spawned `Task` that is simply not `await`ed, so that it runs in parallel to other "logical" threads.

For the current Discord.NET version, we do this with several pieces fitting together:
https://github.com/discord-csharp/MODiX/blob/master/Modix.Common/Messaging/MessageDispatcher.cs
https://github.com/discord-csharp/MODiX/blob/master/Modix.Common/Messaging/INotificationHandler.cs
https://github.com/discord-csharp/MODiX/blob/master/Modix.Services/Core/DiscordSocketListeningBehavior.cs

You can see how subscribing to each event we need, wrapping up the args into an object, and then handing off the work to the dispatcher is a tad tedious.

An `IDiscordEventDispatcher` could look like the following...

```cs
public interface IDiscordEventDispatcher
{
    Task DispatchEvent(IEnumerable<Func<Task>> eventHandlers);
}
```

Where you could have the following built-in implementations that provide behavior similar to `Discord.Commands.RunMode`.

```cs
public class SyncDiscordEventDispatcher
    : IDiscordEventDispatcher
{
    public Task DispatchEvent(IEnumerable<Func<Action>> eventHandlers)
    {
        foreach(var eventHandler in eventHandlers)
            await eventHandler.Invoke();
    }
}
```

```cs
public class ParallelDiscordEventDispatcher
    : IDiscordEventDispatcher
{
    public Task DispatchEvent(IEnumerable<Func<Task>> eventHandlers)
    {
        foreach(var eventHandler in eventHandlers)
            _ = eventAction.Invoke();
        return Task.CompletedTask;
    }
}
```

You can throw in some basic error handling around the `Task`s that aren't `await`ed, or leave that to the consumer.

In our case, we'd implement a custom dispatcher that also handles DI scoping, since Discord.NET doesn't have any built-in DI. Which, actually, now that I think about it, wouldn't fit with the above interface definition.

So, if you're interested at all, it definitely needs some brainstorming. But, I think the concept of handling messages the way ASP.NET handles HTTP requests would be appealing to a lot of consumers.